### PR TITLE
Loss function wrapper to add hydrostatic constraints

### PR DESCRIPTION
**Pull request by akshaysubr**
New loss class that wraps any data loss class (sub-classed from `torch.nn.MSELoss`) and adds hydrostatic constraints

# Modulus Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: physicsnemo/metrics/climate/hydrostasy.py [hunk 1]
```diff
@@ -522,3 +522,312 @@ def forward(self, prediction, target, average_channels=True):
...
+        self.num_z_levels = len(self.z_constraint_pressure_levels)
+        self.num_Tv_levels = len(self.Tv_constraint_pressure_levels)
+        self.z_level_mapping = torch.tensor(list(self.z_pressure_levels.keys()))
+        self.T_level_mapping = torch.tensor(list(self.T_pressure_levels.keys()))
+        self.q_level_mapping = torch.tensor(list(self.q_pressure_levels.keys()))
+
+        # Get topography information
+        ds = xr.open_zarr(f"{src_directory}{dataset_name}.zarr")
+        self.topography = (
+            surface_geopotential_std * ds.constants[1, :, :, :].values
```
Comment (pzharrington) [inline physicsnemo/metrics/climate/hydrostasy.py]: Not the safest to hardcode the orography index in `ds.constants` here -- can we use a safer way of indexing by the name like how land-sea mask is in the other losses? (and throw an error if not found)

File: physicsnemo/metrics/climate/hydrostasy.py [hunk 2]
```diff
@@ -522,3 +522,312 @@ def forward(self, prediction, target, average_channels=True):
...
+                bins=bins if isinstance(bins, int) else bin_edges[l, :],
+            )
+            accumulator[l, :] += hist
+            bin_edges[l, :] = be
+
+        return accumulator, bin_edges
+
+    def forward(self, prediction, target, average_channels=True):
+        """
+        Forward pass of the WeightedMSE pass
```
Comment (pzharrington) [inline physicsnemo/metrics/climate/hydrostasy.py]: Typo -- this is LossWithHydrostasy forward pass

File: physicsnemo/metrics/climate/hydrostasy.py [hunk 3]
```diff
@@ -522,3 +522,312 @@ def forward(self, prediction, target, average_channels=True):
...
+        #Tensors are expected to be in the shape [N, B, F, C, H, W]
+        Tensors are expected to be in the shape [N, F, B, C, H, W]
```
Comment (pzharrington) [inline physicsnemo/metrics/climate/hydrostasy.py:791]: Should remove the commented lines here and below if those are not the expected shapes. In general these shapes need reconciling with the expected shapes in the CRPS loss: https://github.com/AtmosSci-DLESM/modulus-uw/blob/b26a7d47a840e36dbeb17e3a8340a83f75ed3d23/physicsnemo/metrics/climate/healpix_loss.py#L408

In particular there is a lead time dimension `T`, and the CRPS loss assumes the number of conditional ensemble members (`Cond`) has been folded into the batch dimension for the predictions (and is dealt with accordingly).

Comment (pzharrington) [reply physicsnemo/metrics/climate/hydrostasy.py:791]: Since the hydrostatic constraint "broadcasts" across the batch and lead time dimensions, I think the code as you have it here should work, but the docstring should be clearer and more consistent about shapes.

File: physicsnemo/metrics/climate/hydrostasy.py [hunk 4]
```diff
@@ -522,3 +522,312 @@ def forward(self, prediction, target, average_channels=True):
...
+        """
+
+        # Need to scale back to physical units here so disable autocast
+        # and explicitly cast to float32
+        with torch.cuda.amp.autocast(enabled=False):
+            prediction = prediction.float()
+            target = target.float()
+
+            # N, B, F, C, H, W = tuple(prediction.shape)
+            N, F, B, C, H, W = tuple(prediction.shape)
```
Comment (pzharrington) [inline physicsnemo/metrics/climate/hydrostasy.py]: These are unused here so can be dropped

File: physicsnemo/metrics/climate/hydrostasy.py [hunk 5]
```diff
@@ -522,3 +522,312 @@ def forward(self, prediction, target, average_channels=True):
...
+                Tv_error[x[:, :, 1 : self.num_z_levels, :, :] < self.topography] = 0.0
+
+            # Compute the error tolerant loss
+            Tv_loss = self.loss_weights * (Tv_error / (1 + torch.exp(1 - Tv_error))).mean(dim=(0, 1, 3, 4))
+
+            # Compute data loss
+            data_loss = self.data_loss(prediction, target, average_channels=average_channels)
+
+            if average_channels:
+                return data_loss + torch.mean(Tv_loss)
```
Comment (pzharrington) [inline physicsnemo/metrics/climate/hydrostasy.py:828]: Might be convenient to add a scalar weight applied to the net hydrostatic term to help with balancing between different loss scales? Otherwise it would have to be baked into each of the `self.loss_weights` which is more tedious

Comment (akshaysubr) [reply physicsnemo/metrics/climate/hydrostasy.py:828]: That makes sense, but in general, I prefer having config options that are all independent and without redundancy. And either removing one of the loss weights to add an overall loss weighting or somehow rescaling inputs to a non-redundant format seems tedious so I'd prefer leaving it up to the user to set the individual loss weights unless that causes other issues with training experiment setup.

File: physicsnemo/metrics/climate/hydrostasy.py (hunk 4 above)
Comment (akshaysubr) [reply physicsnemo/metrics/climate/hydrostasy.py]: Done

File: physicsnemo/metrics/climate/hydrostasy.py (hunk 3 above)
Comment (akshaysubr) [reply physicsnemo/metrics/climate/hydrostasy.py:791]: Removed the commented line to make it clearer and yes, the way that this is implemented currently works.

File: physicsnemo/metrics/climate/hydrostasy.py (hunk 2 above)
Comment (akshaysubr) [reply physicsnemo/metrics/climate/hydrostasy.py]: Fixed

File: physicsnemo/metrics/climate/hydrostasy.py (hunk 1 above)
Comment (akshaysubr) [reply physicsnemo/metrics/climate/hydrostasy.py]: Fixed by changing to a coordinate based selection of the 'z' constant variable

File: physicsnemo/metrics/climate/hydrostasy.py (hunk 5 above)
Comment (pzharrington) [reply physicsnemo/metrics/climate/hydrostasy.py:828]: Fair enough

**Review by pzharrington — APPROVED** [review]
